### ISSUE TITLE
Save As Copy Does may not correctly deal with Many to Many Fields

### ISSUE THREAD

**Issue by MichaelEPope** [directus/directus#4434]
Hey there, I just wanted to provide some feedback of a bit of unexpected behavior.  I think it should work differently, but perhaps you guys think this is okay behavior?  Let me know what you think.

1) The issue and what you expected to happen
    If you have a Collection with a Many to Many field in it, and you 'Save and Copy' it, the New Record keeps the relationship, but the Old Record loses them (when it should keep them as well).

2) Exact steps to reproduce this issue
    - Create a collection (let's call it A)
    - Create a collection (let's call it B)
    - On the Data Model for Collection A, create a Many-to-Many field in relation to Collection B and a field called 'Name'.
    - On the Data Model for Collection B, crate a Many-to-Many field in relation to Collection A (using the same relationship table as you specified for the previous step) and a field called 'Name'.
    - Create a Record of Type A called C and a record of Type B called D.  Assign B and D to be related using the fields you created in the previous 2 steps (you should only need to assign it in one and it should show up in the other).
    - Save and Copy Record C (you can change the name)
    - Notice that the copy maintains the relationship to Record D.
    - Go back to the original Record C.
    - Notice that the relationship to Record D has been removed.
    - I think that the relationship should have stayed, which means that an extra 'relationship record' needs to be created in the 'relationship table' between the Original Record C and D (or that a new 'relationship record' needs to be created for the New Record C rather than grabbing the one from the Old Record C).

3) Your environment:
    Directus v9.0.04c.40
    Postgres
    Docker
    Brave (Chrome based)

4) Any other relevant information we might need to reproduce this issue
    I don't think so... let me know if you need anything.

**Comment by benhaynes** [comment]
Yeah, you're right! This is something that will likely need a dedicated modal to determine the intended action of the user. In almost all cases, I imagine the user would expect the related data to be duplicated and the new data to be related to the new parent item... but it is likely safest to confirm this with the user, if for now other reason than to make sure they understand what they're doing. So, options in that modal could be:

```
How would you like to handle the related data within "Field Name"?
* Duplicate all child records
* Transfer all child records
* Ignore this field when saving the copy
* Cancel
```

For a M2M, this is controlling the junction items (seems less of a big deal, and MAYBE could be automated), for O2M we're controlling the actual children items so we definitely want to confirm.

**Comment by rijkvanzanten** [comment]
> for O2M we're controlling the actual children items so we definitely want to confirm.

Keep in mind that we can't "duplicate" a O2M structure without copying all related items as well, as the O2M saves straight to the related item in the foreign key

**Comment by benhaynes** [comment]
Yeah, it feels very similar to the "Cascade Delete" issue... but cascade... create?

**Comment by MichaelEPope** [comment]
Yeah, that one to many scenario seems like it'd be a challenge.  Then again... I figure that in those cases, you really want one of two things... you either want a new independent item without any children or you want a duplicate of the children.

In any case, not quite sure about that one.

Hope you are all doing well.

**Comment by MichaelEPope** [comment]
So I think I might have figured out a way to do this... but I can't do it with current functionality.  Is there a way to expose whether or not a copy action is occurring versus a modify action to the items.update.before hook?  Although it might be already exposed... let me check.

**Comment by rijkvanzanten** [comment]
@MichaelEPope There is no dedicated copy action in the api, it's just the app saving the item as new instead of as an update

**Comment by rijkvanzanten** [comment]
There's no state involved in hooks currently, but there shouldn't be anythings topping you from putting a variable in your hook and introduce it 🤔 

```js
module.exports = registerHooks() {
  let state = 0;

  return {
    'items.create': () => { state++ }
  }
}
```

**Comment by MichaelEPope** [comment]
Sweet!  I shall work on that then :D.  Thanks so much

**Comment by rijkvanzanten** [comment]
>   I'd assume in a many to may, what would happen is that the relationship item would receive an update action to update one of it's ids. Kinda like this:
> 
> C -> Rel Record <- D
> C -> Rel Record <- new D and D-> null (the creation of new D updates Rel Record)
> 
> Am I right there? Actually... I should probably just test it. Hmm... I'll have to think about this further.

When cloning a many-to-many, it should inject new rows into the junction table with updated foreign keys that point back to the parent 🤔  .. or is that what you're describing here 🤔 

this gets confusing pretty fast 😅

**Comment by MichaelEPope** [comment]
Huh... what happens to the old rows then?  Do they get deleted?

Currently what I have happening is that either it creates the new rows with the updated foreign keys and deletes the old rows or it just straight up updates the old rows.

Let me expose the primary key of the relationship tables and I'll see which one it does.

Thanks for your patience on this.

**Comment by rijkvanzanten** [comment]
If you're cloning an item, wouldn't you expect the old record to remain unchanged? 🤔  In that case, you'd leave the old junction records as is, and only insert the new records to clone the link on the new item

**Comment by MichaelEPope** [comment]
Yeah... that's the current problem.  Okay, I just tested it and this is what happens in the relationship table.

Basically what I did as an operation is do a 'Save as a Copy' on one of the items in the 'Product' collection.  The Product collection is linked to the 'Hyperlink' collection', so the junction table has a 'product_id', and a 'hyperlink_id'.  When I 'Save as a Copy'd, it actually changed the product_id, rather than creating new lines in the junction table.

Here's the results:


Before
--
id | product_id | hyperlink_id | sort
1 | 13 | 4 | 0
2 | 13 | 5 | 2
3 | 13 | 6 | 4
4 | 13 | 7 | 1
5 | 13 | 8 | 3
6 | 1 | 4 |  
7 | 1 | 7 |  
8 | 1 | 6 |  
9 | 1 | 5 |  
10 | 1 | 8 |  
11 | 10 | 4 |  
12 | 10 | 7 |  
13 | 10 | 6 |  
14 | 10 | 5 |  
15 | 10 | 8 |  
16 | 11 | 4 |  
17 | 11 | 7 |  
18 | 11 | 6 |  
19 | 11 | 5 |  
20 | 11 | 8 |  


After
--
id | product_id | hyperlink_id | sort
1 | 14 | 4 | 0
2 | 14 | 5 | 2
3 | 14 | 6 | 4
4 | 14 | 7 | 1
5 | 14 | 8 | 3
6 | 1 | 4 |  
7 | 1 | 7 |  
8 | 1 | 6 |  
9 | 1 | 5 |  
10 | 1 | 8 |  
11 | 10 | 4 |  
12 | 10 | 7 |  
13 | 10 | 6 |  
14 | 10 | 5 |  
15 | 10 | 8 |  
16 | 11 | 4 |  
17 | 11 | 7 |  
18 | 11 | 6 |  
19 | 11 | 5 |  
20 | 11 | 8 |  

The only changes are the rows with id 1, 2, 3, 4, and 5 changing the product_id from 13 to 14.

Not sure though... this might actually be expected behavior.   If it is, I'll override it with hooks though :).

**Comment by rijkvanzanten** [comment]
Ahh that makes sense.. (from an implementation-perspective, not UX perspective). Right now, it "dumbly" submits to the API what the full item is it received from the API on load, which in turn means that the API will consider the selected rows in the o2m (/m2m) to be marked for update, as it'll update the parent item to be currently created one (it's considered a "select existing" in that case)

---

TL;DR I _think_ the best immediate thing we can do is to strip out relational data, to make sure there's no unexpected updates going on. After that, we should think through a more "wizard" style flow for asking _what_ to copy and how deep to copy it 🤔

**Comment by MichaelEPope** [comment]
That makes sense.  So you guys use the API for the backend too.  Sweet :).

That sounds cool with me.  I'll try and do what I can with a hook in the main time (since I think I can get most of the behavior I want out of it *fingers crossed*)

Thanks for all your help. :)

**Comment by MiniDigger** [comment]
thought about the "wizard" rijk mentioned a bit:

we basically need to have a list of fields that need this special handling. for every of those fields, the user needs asked with of the options they want to choose:
* ignore reference -> keep field in new item empty (default value)
* transfer reference -> move the reference, so the new item has a reference, the old one doesnt
* copy reference -> both items, new and old, reference the same other item
* deepclone reference -> new item references a new item, that has been cloned recursively, by asking these options again 

<details>
<summary>Examples</summary>


Person |   |   |   | Place |   |   |   | City |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
id (PK long) | name (string) | home (FK place) |   | id (PK long) | name (string) | city (FK city) |   | id (PK long) | name (string)
1 | Martin | 1 |   | 1 | Palace | 1 |   | 1 | Cologne
  |   |   |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   |  
UC 1: copy person   martin and rename to rijk, special field detection finds field home, I select   transfer (current behavior)
1 | Martin |   |   | 1 | Palace | 1 |   | 1 | Cologne
2 | Rijk | 1 |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   |  
UC 2: copy person   martin and rename to rijk, special field detection finds field home, I select   ignore (maybe new default, because its the safest?)
1 | Martin | 1 |   | 1 | Palace | 1 |   | 1 | Cologne
2 | Rijk |   |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   |  
UC 3: copy person   martin and rename to rijk, special field detection finds field home, I select   copy reference
1 | Martin | 1 |   | 1 | Palace | 1 |   | 1 | Cologne
2 | Rijk | 1 |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   |  
UC 4: copy person   martin and rename to rijk, special field detection finds field home, I select   deepclone
1 | Martin | 1 |   | 1 | Palace | 1 |   | 1 | Cologne
2 | Rijk | 2 |   | 1 | Palace | ? |   |   |  
  |   |   |   |   |   |   |   |   |  
UC 4.1: now we   need a second step, special field detectin finds city in place palace, prompts me,   I select copy reference
  |   |   |   |   |   |   |   |   |  
1 | Martin | 1 |   | 1 | Palace | 1 |   | 1 | Cologne
2 | Rijk | 2 |   | 1 | Palace | 1 |   |   |  
  |   |   |   |   |   |   |   |   |  
result, both rijk   and martin are different items, that live in difference palaces but in the   same city. User can then, after cloning completes, go to the place and rename   it to Skyscraper

</details>

UI wise this could be a flat list of fields, prefixed by their collection, with a dropdown, which defaults to ignore. whenever I change one to deepclone, directus checks the referenced collection and adds new items to the bottom of the list. recursion can't happen, because each field can only be once in the list.

its a friday afternoon after a short night, but in my head this could work. what do you guys think?

**Comment by MichaelEPope** [comment]
Hey everyone,

Here's a stop-gap I programmed in the main time, in case anybody absolutely needs this functionality right away.  It's really privative, as it uses polling from an external program, but until I get more time, I think it does the trick (yeah, webhooks would be cooler though).  It has a few caveats though (see the top of the code).

Hope your all having an awesome day.

```js
//1)  This only works for many-to-many relationships
//2)  I haven't checked how it works with sorting
//3)  This only works while this program is running (if this program goes down, your many-to-many fix won't work)

const axios = require('axios');               //for making HTTP requests
const async_tools = require('async');         //for helpful asynchronous process control

//from here:  https://stackoverflow.com/questions/22266826/how-can-i-do-a-shallow-comparison-of-the-properties-of-two-objects-with-javascri
function areEqualShallow(a, b) {
    for(var key in a) {
        if(!(key in b) || a[key] !== b[key]) {
            return false;
        }
    }
    for(var key in b) {
        if(!(key in a) || a[key] !== b[key]) {
            return false;
        }
    }
    return true;
}

//data we use to make requests to our Directus API
var base_url = 'PUT URL HERE';
var access_token = 'PUT KEY HERE';

//a function to cause a delay
function delay(time) {
    return new Promise(function(resolve)
      {
      setTimeout(function()
      {
        resolve()
      }, time)
    })
}

//A helper function which gets relations and items (which we will use in the functin below this one)
async function getRelationsAndItems()
{
  //grab the relations from the database
  var relations = await axios.get(base_url + '/relations/', { params: {access_token: access_token}})

  //we need to filter the relations to be many to many and make them unique (so we don't make requests for the same data twice)
  var mtom_uniquify = {};
  relations = relations.data.data.filter(function(relation)
  {
    //if any of these fields are not set, it's not a relation we need to be oncerned with
    if(relation.many_collection == null || relation.many_field == null || relation.junction_field == null)
    {
      return false;
    }
    //for relations we already know we need to grab the data from, we can filter them out
    else if(mtom_uniquify[relation.many_collection])
    {
      return false;
    }
    //otherwise we need to grab the data
    else
    {
      return true;
    }
  })

  //a variable we will stick our relations and items in
  var relation_items = {};

  for(var relation_index = 0; relation_index < relations.length; relation_index++)
  {
    var relation = relations[relation_index]

    //and try to grab the item data for each relation
    try
    {
      var item_list = await axios.get(base_url + '/items/' + relation.many_collection, { params: {access_token: access_token, fields: "*"}});
      item_list = item_list.data.data

      //initialize it in the state object
      relation_items[relation.many_collection] = {};

      //and fill it with the item data
      item_list.forEach(function(data)
      {
        relation_items[relation.many_collection][data.id] = data;
      });
    }
    catch(err)
    {
    }
  }

  //return the data we collected about relationships and items
  return relation_items;
}

var mtom_data;      //We need to save state in between intervals to help with many-to-many relationships (this is the object for that)
var previous_id;    //We need to save state (in this case, the activity id) to see if an update has occured in Directus
async_tools.forever(async function()
{
  //first off, we need to grab the activity (so we can see if we need to restart the server later)
  var activity = await axios.get(base_url + '/activity/', { params: {access_token: access_token, limit: 1, sort: "-id"}})
  activity = activity.data.data[0];

  //if there wasn't an activity id in our state object, fill it in
  if(!previous_id)
  {
    previous_id = activity.id;
  }

  //if there isn't any mtom_data in our state object, fill it in
  if(!mtom_data)
  {
    mtom_data = await getRelationsAndItems();
  }

  //otherwise if the activity has changed
  else if(previous_id != activity.id)
  {
    //grab a list of new data
    var new_mtom_data = await getRelationsAndItems();

    //for each collection
    for(collection in new_mtom_data)
    {
      //get the old and new data
      var new_data = new_mtom_data[collection];
      var old_data = mtom_data[collection];

      //and compare each id (we are just looking for modifications, (so created and deleted don't matter))
      if(old_data)
      {
        for(id in new_data)
        {
          if(old_data[id])
          {
            //if they are not equal, duplicate the old object back to Directus
            if(!areEqualShallow(old_data[id], new_data[id]))
            {
              var copy = old_data[id];
              delete copy.id;
              await axios.post(base_url + '/items/' + collection, copy, { params: {access_token: access_token}});
            }
          }
        }
      }
    }
    mtom_data = new_mtom_data;

    //if you are restarting a static web server because of changes, this would be the place to trigger it

  }

  //make sure there is at least 1.5 seconds between polls
  await delay(1500);
  return;
})
```

**Comment by rijkvanzanten** [comment]
OK! After some more thinking, here's a possible continuation of this discussion, based on [@MiniDigger's super helpful comment](https://github.com/directus/directus/issues/4434#issuecomment-797474048) detailing some use cases. (Warning: this is a bit of a brain-fart ramble)

I believe this whole discussion should only apply to _one to many_ relationships. I know the OP mentions `m2m`, but in the context of Directus, that's nothing more than a `o2m-m2o` combination. Many to One keys are stored in a column in the parent, and can therefore be copied just like any other field without causing any trouble elsewhere. (Btw @MiniDigger in your example above, from Person to Place, that FK reference is a m2o, and can thus "just" be copied without any side effects right? 🤔 )

Next up: I'm wondering if this should be a case-by-case, or if it's enough to "package" this in with #3719 and make it a field option ("On clone/update, do [      ]. On delete, do [       ]"). That would take away a lot of end-user confusion (what does it mean to "deep clone" etc)

Taking this as a starting point:

> ignore reference -> keep field in new item empty (default value)
> transfer reference -> move the reference, so the new item has a reference, the old one doesnt
> copy reference -> both items, new and old, reference the same other item
> deepclone reference -> new item references a new item, that has been cloned recursively, by asking these options again

ignoring sounds like a good default to me as well, as it's the path of "least unexpected side effects". 
I do think "transfer" should be left out in the context of "Copy". IMO we should avoid modifying the "to be copied" item at any costs, as that feels like an unexpected side effect to me "Wait I _copied_ the item, why did the original change?"

At first, I was a little hesitant when thinking through "deep clone", as that has the potential for an infinite depth of recursive paths (what if the m2m goes both ways or what if you have articles -> categories (user created) -> users (favorite articles) -> articles 🔁 ). However, hopefully this isn't as much of a problem we're only concerned about One to Many relations 🤔  This also ties back in with the question: case-by-case or on the field level? By configuring it in the field, we don't end up having to ask the user over and over for nested fields

---

TL;DR:

* Only consider one-to-many's, as m2o's are a value that can be copied
* Never alter the to-be-copied item, as that would go against the expected flow of "copying" something
* Consider saving this as a configuration on the field level, together with recursive delete #3719
* Need more research into circular o2m-structures (do those even exist IRL?)

---

(@MiniDigger I believe your example would be correct if it was `Person` -> `Person_Place` -> `Place` where `Person` has a `favorite_places` alias o2m to `Person_Place`, with `Person_Place` having a FK to both `Person` and `Place` (eg it being a m2m junction table))

**Comment by MiniDigger** [comment]
I like the idea of configuring that on the field itself, that makes it way easier for users.
I agree that "transfer" is confusing and I dont think there is a need for that.
I think my example was so flawed, because I didnt consider how directus models relations. you looked at it from the DB side and I looked at it more from "how do the collections look like in directus" and thus missed the junction table (smth smth I blame friday afternoon)

this brings me to my real world example, which features a m2a, would that work too?

basically, I have a page collection, which has an m2m to paragraph (so I got a page_paragraph junction table). a paragraph then has a m2a to elements (text element, picture element, etc). so theres a junction table again. 
am I right that this comes down to the m2o again?

the user here would expect that when they clone the page, all paragrahs and all elements in those paragraphs gets cloned to, so that it is save to change text in a text element in a paragraph, without modifying other pages elements.

one more important thing is to keep in mind that junction tables can have more than just FKs, so their other fields need to be copied too.

**Comment by rijkvanzanten** [comment]
> the user here would expect that when they clone the page, all paragrahs and all elements in those paragraphs gets cloned to, so that it is save to change text in a text element in a paragraph, without modifying other pages elements.

This is a very interesting point. In the flow I mentioned above, the row on the junction table "pages_sections" would stay the same, but the related m2a fk to the actual paragraph would be duplicated, meaning that the new page would have a reference to the _same_ paragraph (similar as using the "Select Existing" button in the interface instead of "Add New") 🤔

**Comment by MichaelEPope** [comment]
Hey @rijkvanzanten,
By 'copy reference -> both items, new and old, reference the same other item', you mean that both 'many' items (the original and the copy) would reference the 'one' item, right?

Hope your having an awesome day.

**Comment by rijkvanzanten** [comment]
@MichaelEPope 

Lets say you have "Articles" and "Categories". An article can have multiple categories, and likewise, a category can have multiple associated articles. You'd set this up as follows:

```
articles    articles_categories    categories
--------    -------------------    ----------
id          id                     id
title       article_id             name
body        category_id
```

Now, if you're looking at it from an article detail page (so a single article), you'd have a list of categories in a m2m, but the list you're looking at / modifying is secretly the list of "articles_categories" rows. This means that "add new" is adding a new junction row _and_ a new "categories" row, and "select existing" is only creating a new junction row.

With this example in mind, the question becomes: "What do you expect to happen when you clone the article?"

What @MiniDigger meant by `'copy reference -> both items, new and old, reference the same other item` is that in case of a clone of the `articles` row, we'd duplicate all rows in `articles_categories` where `article_id` = current article ID, update `article_id` to the newly created article, but _leave_ the `category_id` column the same. This way, the new article has all the same categories as the old article.

This is the same general idea that I meant when saying we'd "only follow the o2m side" as articles->articles_categories is a o2m (so gets "cloned" as well), but the articles_categories->categories is a m2o, so will be left alone.

_However_, there is a case to be made where the item on "the other end" needs to be duplicated as well! In @MiniDigger's example above, we're talking about a pages -> pages_sections -> sections case, where the section is ideally duplicated as well......

(That being said, I'm wondering if that is a problem with the fact that the current m2a approach is structured as a m2m instead of as a o2m)

**Comment by MichaelEPope** [comment]
That makes total sense.  Cool :).

Hmmm... so I may be going in the wrong direction here, but I think you actually made the right approach regarding the m2a fields as they are now.  While most cases would probably work best with a "o2a" field, I think "m2a" catches some cases that "o2a" wouldn't.  As an example let's say you have [PRODUCTS}, and your trying to build up a relation to [SOFTWARE] and [FIRMWARE] collections.  The software and firmware apply to multiple products.  While you could have 2 "m2m" field, a "m2a" works pretty well.

I create a m2a field and the structure, as you said, seems pretty similar to the m2m.  Your basically just adding a collection name field, to the junction table.  So maybe I'm missing things here... but I'm not imagining that 'deep copying' a m2a would be much more difficult than deep copying a m2m.  I might be wrong there though.  As you've mentioned before, circular references are going to be a pain though.

Hope your having a good day.

**Comment by rijkvanzanten** [comment]
That's very true too! So now we have a problem: you might want to copy _some_ of the related items, but not all. I also recognize the use case that you might want to duplicate the related paragraph section on a page, but not the related product (as that's re-used across multiple pages). Difficult difficult!

> but I'm not imagining that 'deep copying' a m2a would be much more difficult than deep copying a m2m

This is true, they're effectively the same when discussing this deep-cloning

> I think the main difficulty is probably about the difference between 'relationship' (see the products, firmware, software example) and 'ownership' (see the pages, page section, section case example), and m2m's and m2a's can be used in both ways.

This is a perfect way to describe the difference between these two! I'm just really worried about the technical (im-)possibilities of trying to make this work for these both.. Like you mentioned, once we start cloning the "m2o" side of these structures too, you end up with circular dependencies _fast_ 😱 

> Hope your having a good day, and sorry if this is being a hassle at all.

Same! ❤️  No hassle at all 🙂 As a matter of fact, I'm super glad you're thinking through this with me, cause it's super complex to figure out the ideal scenario for all this

**Comment by stouch** [comment]
Just to be sure : I encounter a problem when I _save as copy_ an item of a collection A, which has a O2M field `related_b_items` that points to several items of collection B : My item of A is copied (a -> a.bis), but the related items `related_b_items` are not copied, just moved (their foreign key) to a.bis .

Is this an expected behavior ?
